src/rust/engine/process_execution/src/nailgun/nailgun_pool.rs

```diff
@@ -50,124 +99,134 @@ impl NailgunPool {
                  Startup Options: {:?}\n Process Cmd: {:?}",
-                &process_name, startup_options, process_fingerprint
-              );
-              debug!("Restarting the server...");
-              self.start_new_nailgun(
-                &mut *processes,
-                name,
-                startup_options,
-                workdir_path,
-                requested_server_fingerprint,
-                build_id_requesting_connection,
-              )
+                  &process_name, startup_options, process_fingerprint
+                );
+                debug!("Restarting the server...");
+                Self::start_new_nailgun(
+                  &mut *processes,
+                  name,
+                  startup_options,
+                  &workdir_path,
+                  requested_server_fingerprint,
+                  build_id_requesting_connection,
+                )
+              }
             }
           }
+          Ok(_) => {
+            // The process has exited with some exit code
+            debug!("The requested nailgun server was not running anymore. Restarting process...");
+            Self::start_new_nailgun(
+              &mut *processes,
+              name,
+              startup_options,
+              &workdir_path,
+              requested_server_fingerprint,
+              build_id_requesting_connection,
+            )
+          }
+          Err(e) => Err(e),
         }
-        Ok(_) => {
-          // The process has exited with some exit code
-          debug!("The requested nailgun server was not running anymore. Restarting process...");
-          self.start_new_nailgun(
-            &mut *processes,
-            name,
-            startup_options,
-            workdir_path,
-            requested_server_fingerprint,
-            build_id_requesting_connection,
-          )
-        }
-        Err(e) => Err(e),
-      }
-    } else {
-      // We don't have a running nailgun registered in the map.
-      debug!(
-        "No nailgun server is running with name {}. Starting one...",
-        &name
-      );
-      self.start_new_nailgun(
-        &mut *processes,
-        name,
-        startup_options,
-        workdir_path,
-        requested_server_fingerprint,
-        build_id_requesting_connection,
-      )
-    };
-    debug!("Unlocking nailgun process pool.");
-    connection_result
+      } else {
+        // We don't have a running nailgun registered in the map.
+        debug!(
+          "No nailgun server is running with name {}. Starting one...",
+          &name
+        );
+        Self::start_new_nailgun(
+          &mut *processes,
+          name,
+          startup_options,
+          &workdir_path,
+          requested_server_fingerprint,
+          build_id_requesting_connection,
+        )
+      };
+      debug!("Unlocking nailgun process pool.");
+      connection_result
+    })
+    .to_boxed()
   }
 
   fn start_new_nailgun(
```

Comment:
Would you mind adding a comment here indicating that this is a blocking method that is being called under the `NailgunProcessMap`'s lock, and referring to #8481? I'll add a comment there to suggest fixing this later.